stdlib/public/runtime/HeapObject.cpp

```diff
@@ -625,18 +593,28 @@ void swift::swift_rootObjCDealloc(HeapObject *self) {
 }
 #endif
 
+#if SWIFT_OBJC_INTEROP
+static void _check_fast_dealloc(void *ctxt) {
+  auto answer = reinterpret_cast<bool *>(ctxt);
+  //This will always be in libobjc, so RTLD_DEFAULT won't have to do an
+  //expensive search in practice
+  *answer = dlsym(RTLD_DEFAULT, "_objc_has_weak_formation_callout") != nullptr;
+}
+#endif
+
 void swift::swift_deallocClassInstance(HeapObject *object,
                                        size_t allocatedSize,
                                        size_t allocatedAlignMask) {
-#if OBJC_SETASSOCIATEDOBJECTHOOK_DEFINED
+  
+#if SWIFT_OBJC_INTEROP
   // We need to let the ObjC runtime clean up any associated objects or weak
   // references associated with this object.
-  if (originalAssocObjectFunc == nullptr ||
-      !object->refCounts.getPureSwiftDeallocation()) {
+  static bool fastDeallocSupported;
+  static swift_once_t fastDeallocOnce;
+  swift_once(&fastDeallocOnce, _check_fast_dealloc, &fastDeallocSupported);
```

Comment:
You might use `SWIFT_LAZY_CONSTANT` to cut down on some boilerplate here. We're not 100% consistent but that appears to be the preferred way to do these things in the codebase. I think/hope it compiles down to the same stuff.

Reply:
Ah good call